### PR TITLE
PercentOf: fix #1109 bug

### LINKED ISSUE #1109: Percent calculation looks wrong
cfdwcdz
1tan1jp

https://twitter.com/dsokol/status/590150769103474689?refsrc=email&s=11

### PR DESCRIPTION
**Pull request by puskin94**
Hei, I fixed this bug: https://github.com/duckduckgo/zeroclickinfo-goodies/issues/1109

Sorry for the drawback ! :)

// cc @yegg @MrChrisW

### CODE REVIEW

**Comment by MrChrisW** [comment]
@puskin94 Thanks for the quick fix :+1: 
Do you mind updating the test file to include the queries mentioned in the issue report?

**Comment by puskin94** [comment]
@MrChr No, it was only bug in calculations! Queries are the same ;) I think it can be good now!

**Comment by mintsoft** [comment]
@puskin94 please add a test for the query that caused the error with the expected output. The idea is that the tests shape the desired behaviour of the Goodies. If a problem is found with the output then a test should be added to show people cleaning up or maybe merging the IA into other IA's what users expect.

**Comment by puskin94** [comment]
@mintsoft @MrChrisW Ok! Updated test file too ;)
Thanks for your time :+1:

**Comment by MrChrisW** [comment]
Thanks @puskin94, while we're here I just noticed this query results in an incorrect answer, Not sure if this is an easy fix. 

It looks like it's subtracting the percent from the last number, I think it should perform the subtraction on the sum of the first two? @mintsoft opinion? 


selection_385

**Comment by puskin94** [comment]
@MrChrisW yep, In the first PR there war the doubt about merging this IA into Calculator.pm but it is difficult to debug/modify. This IA makes a simple calculation on two terms

**Comment by MrChrisW** [comment]
@puskin94 Sure understood -- maybe we should avoid showing a result for these types of queries? 

Have you seen the new JS calculator [1727](https://github.com/duckduckgo/zeroclickinfo-spice/pull/1727) :tada:

**Comment by puskin94** [comment]
@MrChrisW Yep! It is really cool and it has a very good GUI :) I'm working on a new Regex that can handle all this cases! This evening you should see a new commit ;)

**Comment by MrChrisW** [comment]
>I'm working on a new Regex that can handle all this cases! This evening you should see a new commit ;)

@puskin94 No problem! take your time, as always feel free to ping when ready :smiley:

**Comment by puskin94** [comment]
@MrChrisW What about now? Just tell me new possible changes ;) Thanks for your interest in this IA!

**Comment by puskin94** [comment]
0.o This is new for me! Thanks for the tip ;) Do you know why tests are failing ??

**Comment by MrChrisW** [comment]
@puskin94 Do the tests pass locally? Currently tests are failing for other reasons (related to `[AutoPrereqs]`)
https://github.com/duckduckgo/zeroclickinfo-spice/issues/1805

**Comment by puskin94** [comment]
@MrChrisW Yep, All the tests are passed locally

**Comment by MrChrisW** [comment]
Thanks @puskin94 assigning to @mintsoft for review.

File: lib/DDG/Goodie/PercentOf.pm [hunk 1]
```diff
@@ -20,25 +20,25 @@ triggers query_nowhitespace => qr/\d{1,3}\%$/;
...
+        $result = ( $1 - ( $partRes ) );
     } elsif ($2 eq '+') {
-        $result = ( $1 + (($1 * $3) / 100) );
+        $result = ( $1 + ( $partRes ) );
     } elsif ($2 eq '*') {
-        $result = ( $1 * (($1 * $3) / 100) );
+        $result = ( $partRes );
     } elsif ($2 eq '/') {
-        $result = ( $1 / (($1 * $3) / 100) );
+        $result = ( $1 * 100 / ( $partRes ) );
```
Comment (mintsoft) [inline lib/DDG/Goodie/PercentOf.pm]: Can you add a test for the `/` case; it looks like the behavior was changed here. If it was incorrect previously can you add another test for the correct output?

File: t/PercentOf.t [hunk 2]
```diff
@@ -13,29 +13,54 @@ ddg_goodie_test(
...
-            input => ["3.4+6%"],
+            input => ['3.4+6%'],
             operation => "Calculate",
             result => 3.604
         }),
+
     '323.7+55.3%' => test_zci('Result: 502.7061',
         structured_answer => {
-            input => ["323.7+55.3%"],
+            input => ['323.7+55.3%'],
```
Comment (mintsoft) [inline t/PercentOf.t]: Also there seems to be a lack of tests around queries such as `123 + 3%` or `123+ 5%` or `123 +2%`

**Comment by puskin94** [comment]
@mintsoft Done! Thanks for this review!

**Comment by yegg** [comment]
Thanks all!

**Comment by mintsoft** [comment]
Thanks for the fix @puskin94 !

**Comment by puskin94** [comment]
@mintsoft @yegg @MrChrisW thanks you all for your time :+1:

**Pull request merged**